kv: do not fatal when assertRefreshSpansAtInvalidTimestamp errors

arulajmani opened the pull request:
The Txn API prohibits using the root transaction when one or more leaf transactions that have been forked from it are still active. One reason for this is that an operation on a root transaction can cause it to perform a read refresh; however, leaf transactions cannot perform read refreshes and we have no way to talk about a "partially refreshed" set of read spans. As such, when ingesting state from a leaf transaction, the root calls into
 `assertRefreshSpansAtInvalidTimestamp` which makes sure the timestamp
at which the leaf performed its reads is not below the timestamp at which the root has already been refreshed to.

Previously, tripping this assertion would cause the gateway node to crash because we'd log this error as a Fatal. Even though this situation results from a programming mistake by users of the Txn API, the correctness issue is isolated to the query in question. It doesn't cascade to other queries/cause data corruption. As such, it shouldn't warrant crashing the node. Instead, it should be sufficient to simply return this assertion failed error back to the client. While this situation is

Resolves #99255

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99306)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 13 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99306)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`-- commits` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-6GK2JMdYOFW5dF-):*
While this situation is what?
___
*[`-- commits` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-AYo34kqlF9mUsPj):*
nit: I don't think we need this if we have "Resolves ..."
___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 1232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-TM789dCn9N2QvQj:-NRB-TM789dCn9N2QvQk:b-2cnnj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/38d13e02c5efbc74577f5246a94686494a4d9540/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L1232)):*
> ```Go
> 	// Sanity check: don't combine if the tfs is for a different txn ID.
> 	if tc.mu.txn.ID != tfs.Txn.ID {
> 		// TODO(arul): This should be an AssertionFailed error.
> ```

Want to do that now while we're here?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go` line 641 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-kdnBjTyue7-fXPj:-NRB-kdnBjTyue7-fXPk:b-ctldjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/38d13e02c5efbc74577f5246a94686494a4d9540/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L641)):*
> ```Go
> }
> 
> const RefreshSpansAtInvalidTimestampMsg = "attempting to append refresh spans after the tracked " +
> ```

We generally err on the side of a bit of duplication and imprecision when asserting error strings instead of leaking testing-only error strings into each package's exported namespace. So feel free to just inline this again and use something like the following in your test:
```go
require.True(t, errors.IsAssertionFailure(err), "%+v", err)
require.Regexp(t, "attempting to append refresh spans after the tracked timestamp has moved forward", err)
```
___
*[`pkg/kv/txn_external_test.go` line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-wDzBf-OVyhU4Wx3:-NRB-wDzBf-OVyhU4Wx4:bvjp7y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/38d13e02c5efbc74577f5246a94686494a4d9540/pkg/kv/txn_external_test.go#L831)):*
> ```Go
> 		err = txn.UpdateRootWithLeafFinalState(ctx, finalState)
> 		require.Error(t, err)
> 		testutils.IsError(
> ```

`testutils.IsError` returns a bool, so this isn't actually asserting anything. We've moved away from this in favor of `require.Regexp`, so consider switching to that.


<!-- Sent from Reviewable.io -->

arulajmani committed:
kv: do not fatal when assertRefreshSpansAtInvalidTimestamp errors

The Txn API prohibits using the root transaction when one or more
leaf transactions that have been forked from it are still active. One
reason for this is that an operation on a root transaction can cause it
to perform a read refresh; however, leaf transactions cannot perform
read refreshes and we have no way to talk about a "partially refreshed"
set of read spans. As such, when ingesting state from a leaf
transaction, the root calls into
`assertRefreshSpansAtInvalidTimestamp` which makes sure the timestamp
at which the leaf performed its reads is not below the timestamp at
which the root has already been refreshed to.

Previously, tripping this assertion would cause the gateway node to
crash because we'd log this error as a Fatal. Even though this situation
results from a programming mistake by users of the Txn API, the
correctness issue is isolated to the query in question. It doesn't
cascade to other queries/cause data corruption. As such, it shouldn't
warrant crashing the node. Instead, it should be sufficient to simply
return this assertion failed error back to the client.

Resolves #99255

Release note: None

arulajmani reviewed (commented):
RFAL. 

----

@yuzefovich:

> How can I force the read refresh of the RootTxn in this scenario? I need concurrent writes to system.jobs?

Yeah. I think in this construction we need the write to happen at a timestamp at a timestamp above the RootTxn's read timestamp, but within the uncertainty interval. That's how we'll be able to force a read refresh on the RootTxn that's only doing reads. 

> Reads via the LeafTxn in the demo setup are always performed after all reads in the RootTxn. Do I need them to be intertwined (i.e. I need the lookup join to perform multiple batches of reads, which would mean I need to more rows in crdb_internal.jobs table)?

Technically, all we need is for the LeafTxn to be initialized before the RootTxn is refreshed, so that it gets the timestamp below the refresh. I could be wrong, but I _think_ that means you should be able to get by without needing multiple batches of reads?  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-:-NRB5PEr7apxYikAnM7t:batf34k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`-- commits` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-6GK2JMdYOFW5dF-):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

While this situation is what?
</blockquote></details>

detritus from a try to reword this paragraph. Removed. 
___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 1232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-TM789dCn9N2QvQj:-NRB5AWHAXnDxEACfcAR:bb8dqyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/38d13e02c5efbc74577f5246a94686494a4d9540/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L1232)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Want to do that now while we're here?
</blockquote></details>

Yup, done. 
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go` line 641 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-kdnBjTyue7-fXPj:-NRB4QRG3oVhzNKMwAEV:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/38d13e02c5efbc74577f5246a94686494a4d9540/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L641)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We generally err on the side of a bit of duplication and imprecision when asserting error strings instead of leaking testing-only error strings into each package's exported namespace. So feel free to just inline this again and use something like the following in your test:
```go
require.True(t, errors.IsAssertionFailure(err), "%+v", err)
require.Regexp(t, "attempting to append refresh spans after the tracked timestamp has moved forward", err)
```
</blockquote></details>

Done. 
___
*[`pkg/kv/txn_external_test.go` line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-NRB-wDzBf-OVyhU4Wx3:-NRB4RzRByS_k_ejGOap:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/38d13e02c5efbc74577f5246a94686494a4d9540/pkg/kv/txn_external_test.go#L831)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`testutils.IsError` returns a bool, so this isn't actually asserting anything. We've moved away from this in favor of `require.Regexp`, so consider switching to that.
</blockquote></details>

Done. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99306#-:-NRB8hSx0Gf8-uAfKgVP:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)



<!-- Sent from Reviewable.io -->
